privatekey/privatekey.go

```diff
@@ -0,0 +1,113 @@
+package privatekey
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"hash"
+	"io/ioutil"
+)
+
+// verifyRSA is broken out of Verify for testing purposes.
+func verifyRSA(privKey *rsa.PrivateKey, pubKey *rsa.PublicKey, msgHash hash.Hash) error {
+	signatureRSA, err := rsa.SignPSS(rand.Reader, privKey, crypto.SHA256, msgHash.Sum(nil), nil)
+	if err != nil {
+		return fmt.Errorf("failed to sign using the provided RSA private key: %s", err)
+	}
+
+	err = rsa.VerifyPSS(pubKey, crypto.SHA256, msgHash.Sum(nil), signatureRSA, nil)
+	if err != nil {
+		return fmt.Errorf("the provided RSA private key failed signature verification: %s", err)
+	}
+	return err
+}
+
+// verifyECDSA is broken out of Verify for testing purposes.
+func verifyECDSA(privKey *ecdsa.PrivateKey, pubKey *ecdsa.PublicKey, msgHash hash.Hash) error {
+	r, s, err := ecdsa.Sign(rand.Reader, privKey, msgHash.Sum(nil))
+	if err != nil {
+		return fmt.Errorf("failed to sign using the provided ECDSA private key: %s", err)
+	}
+
+	verify := ecdsa.Verify(pubKey, msgHash.Sum(nil), r, s)
+	if !verify {
+		return errors.New("the provided ECDSA private key failed signature verification")
+	}
+	return err
+}
+
+// Verify ensures that the embedded PublicKey of the provided privateKey is
+// actually a match for the private key. For an example of private keys
+// embedding a mismatched public key, see:
+// https://blog.hboeck.de/archives/888-How-I-tricked-Symantec-with-a-Fake-Private-Key.html.
+func Verify(privateKey crypto.Signer) error {
+	msgHash := sha256.New()
+	_, err := msgHash.Write([]byte("verifiable"))
+	if err != nil {
+		return fmt.Errorf("failed to hash 'verifiable' message: %s", err)
+	}
+
+	switch k := privateKey.(type) {
+	case *rsa.PrivateKey:
+		return verifyRSA(k, &k.PublicKey, msgHash)
+
+	case *ecdsa.PrivateKey:
+		return verifyECDSA(k, &k.PublicKey, msgHash)
+
+	default:
+		// This should never happen.
+		return errors.New("the provided private key could not be asserted to ECDSA or RSA")
+	}
+}
+
+// Load decodes and parses a private key from the provided file path and returns
+// the private key as crypto.Signer. path is expected to be a PEM formatted RSA
+// or ECDSA private key in a PKCS #1, PKCS# 8, or SEC 1 container.
```

Comment:
I think it's worth noting that this function should only be used in test, one-off, non-prod, exceptional, etc. cases. In general, Boulder only wants to be talking to HSMs, not loading private keys into memory. We should be careful with when and where we use this function.